File: src/61_Rotate_List.rs
// 61. Rotate List
// https://leetcode.cn/problems/rotate-list/

// Definition for singly-linked list.

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
        next: None,
        val
        }
    }
}


impl Solution {
    pub fn rotate_right(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        if head.is_none() || k == 0 {
            return head;
        }

        let mut head = head;
        let mut ptr = &head;
        let mut len = 0;
        while let Some(ref t) = ptr {
            ptr = &t.next;
            len += 1;
        }

        let k = k % len;
        if k == 0 {
            return head;
        }

        let mut ptr = &mut head;

        for _ in 1..len - k {
            ptr = &mut ptr.as_mut().unwrap().next;
        }

        let mut newhead = ptr.as_mut().unwrap().next.take();
        let mut tail = &mut newhead;
        while tail.is_some() &&  tail.as_ref().unwrap().next.is_some() {
            tail = &mut tail.as_mut().unwrap().next;
        }
        tail.as_mut().unwrap().next = head;
        newhead
    }
}

File: src/61_Rotate_List.c
// https://leetcode.cn/problems/rotate-list/
// 4 ms 71.43%
// 6 MB 59.61%

#include <stdio.h>

// Definition for singly-linked list.
struct ListNode
{
    int val;
    struct ListNode *next;
};

// struct ListNode* rotateRight(struct ListNode* head, int k){
//     struct ListNode dummy;
//     struct ListNode *tmp = head;

//     int len = get_length(head);
//     if (len == 0) {
//         return head;
//     }
//     k = k % len;
//     for (int i=0; i < len-k; i++) {
//         tmp = tmp->next;
//     }
//     dummy.next = tmp;
//     while (tmp) {
//         tmp = tmp->next;
//     }
//     tmp->next = head;
//     for (int i=0; i<len-k-1; i++) {
//         head = head->next;
//     }
//     head->next = NULL;

//     return dummy.next;
// }

struct ListNode *rotateRight(struct ListNode *head, int k)
{
    if (!head || !k)
        return head;
    int len = 0;
    struct ListNode *tail;
    for (struct ListNode *pre = head; pre; pre = pre->next)
    {
        tail = pre;
        ++len;
    }
    k %= len;
    struct ListNode *pre = head;
    for (int i = 0; i < len - k - 1; i++)
    {
        pre = pre->next;
    } // 找到链表的第n-k个节点
    tail->next = head;
    head = pre->next;
    pre->next = NULL;

    return head;
}
